### PR TITLE
Add support for 'named tensors'

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
Title says it all.

See: https://pytorch.org/docs/stable/named_tensor.html?highlight=named%20tensors

I also renamed 'requiresGrad' to 'requires_grad' everywhere it was found as a parameter name.

### CODE REVIEW

**Comment by NiklasGustafsson** [comment]
@MovGP0 -- this is in response to your issue #568. If you have a change to take a look, that would be great.

File: src/Native/LibTorchSharp/THSTensor.cpp [hunk 1]
```diff
@@ -48,7 +48,7 @@ Tensor THSTensor_aminmax(const Tensor tensor, const int64_t dim, bool keepdim, T
 {
     std::tuple<at::Tensor, at::Tensor> res;
     CATCH(res = dim == -1 ? tensor->aminmax(c10::nullopt, keepdim) : tensor->aminmax(dim, keepdim);)
-    *max = ResultTensor(std::get<1>(res));
+        * max = ResultTensor(std::get<1>(res));
```
Comment (tarekgh) [inline src/Native/LibTorchSharp/THSTensor.cpp:51]: >  [](http://example.com/codeflow?start=9&length=1)

nit: remove this space.

Comment (NiklasGustafsson) [reply src/Native/LibTorchSharp/THSTensor.cpp:51]: Sure. BTW, Ctrl-K,Ctrl-D does such a poor job in these C++ files with the CATCH() macro, that I have generally given up on the indentation. This isn't one of those, I don't think, but I thought I'd mention it.

File: src/TorchSharp/Tensor/Tensor.cs [hunk 2]
```diff
@@ -881,6 +883,214 @@ public long[] size()
...
+                IntPtr res = IntPtr.Zero;
+
+                if (names != null && names.Count() > 0) {
+
+                    var dimNamesArray = ExpandEllipsis(names);
+
+                    IntPtr namesRef = IntPtr.Zero;
+
+                    using PinnedArray<IntPtr> pinnedArray = new PinnedArray<IntPtr>();
+                    namesRef = (names is null || names.Count() == 0) ? IntPtr.Zero : pinnedArray.CreateArray(dimNamesArray);
```
Comment (tarekgh) [inline src/TorchSharp/Tensor/Tensor.cs:970]: >(names is null || names.Count() == 0) [](http://example.com/codeflow?start=31&length=37)

Is this condition possible to be true?

File: src/TorchSharp/Tensor/Tensor.cs [hunk 3]
```diff
@@ -881,6 +883,214 @@ public long[] size()
...
```
Comment (tarekgh) [inline src/TorchSharp/Tensor/Tensor.cs:1001]: >(names is null || names.Count() == 0) [](http://example.com/codeflow?start=31&length=37)

same as previous comment.

File: src/TorchSharp/Tensor/Tensor.cs [hunk 4]
```diff
@@ -881,6 +883,214 @@ public long[] size()
...
+                IntPtr namesRef = pinnedArray.CreateArray(dimNamesArray);
+
+                IntPtr res = THSTensor_refine_names(Handle, namesRef, dimNamesArray.Length);
+                if (res == IntPtr.Zero) { torch.CheckForErrors(); }
+                return new Tensor(res);
+            }
+
+            private IntPtr[] ExpandEllipsis(IEnumerable<string?> names)
+            {
+                var namesArray = names.ToArray();
```
Comment (tarekgh) [inline src/TorchSharp/Tensor/Tensor.cs:1039]: >names.ToArray(); [](http://example.com/codeflow?start=33&length=16)

do we need this allocation? we can iterate over the IEnumerable. right? Also, IEnumerable support `Contains` method too. Or you think this is not hot path to optimize?

File: src/TorchSharp/Tensor/Tensor.cs (hunk 2 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/Tensor/Tensor.cs:970]: No. Thanks for catching.

File: src/TorchSharp/Tensor/Tensor.cs (hunk 4 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/Tensor/Tensor.cs:1039]: No.

Comment (NiklasGustafsson) [reply src/TorchSharp/Tensor/Tensor.cs:1039]: The thing with this code is that there's a fair amount of indexing of several collections. It seemed easier to use indices everywhere, but I guess namesArray can remain an IEnumerable, actually. The code will be a little less readable, but not by much.

File: src/TorchSharp/Tensor/Tensor.cs [hunk 5]
```diff
@@ -881,6 +883,229 @@ public long[] size()
...
+
+                    if (idx == namesArray.Length - 1) { // '...' appears last.
+
+                        int i = 0;
+                        for (; i < idx; i++)
+                            dimNamesArray[i] = MarshalDimensionString(namesArray[i]);
+
+                        if (has_names()) {
+
+                            var n = this.names.ToArray();
```
Comment (tarekgh) [inline src/TorchSharp/Tensor/Tensor.cs:1055]: >var n = this.names.ToArray(); [](http://example.com/codeflow?start=28&length=29)

we can git id of this allocation too I guess.

File: src/TorchSharp/Tensor/Tensor.cs [hunk 6]
```diff
@@ -881,6 +883,229 @@ public long[] size()
...
+
+                        int dIdx = 0;
+                        for (int i = 0; i < idx; i++, dIdx++)
+                            dimNamesArray[dIdx] = MarshalDimensionString(namesArray[i]);
+
+                        int missing_dims = (int)(ndim - namesArray.Length + 1);
+
+                        if (has_names()) {
+
+                            var n = this.names.ToArray();
```
Comment (tarekgh) [inline src/TorchSharp/Tensor/Tensor.cs:1075]: > var n = this.names.ToArray(); [](http://example.com/codeflow?start=27&length=30)

her too we can git id of this allocation too I guess.

**Review by tarekgh — APPROVED** [review]

File: src/TorchSharp/Tensor/Tensor.cs (hunk 6 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/Tensor/Tensor.cs:1075]: Yup.

File: src/TorchSharp/Tensor/Tensor.cs (hunk 5 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/Tensor/Tensor.cs:1055]: Yes.

**Comment by GeorgeS2019** [comment]
## WIP [NAMED TENSORS](https://pytorch.org/docs/stable/named_tensor.html?highlight=named%20tensors)

### C[reating named tensors](https://pytorch.org/docs/stable/named_tensor.html#creating-named-tensors)
- [ ] [torch.empty()](https://pytorch.org/docs/stable/generated/torch.empty.html#torch.empty)

- [ ] [torch.rand()](https://pytorch.org/docs/stable/generated/torch.rand.html#torch.rand)

- [ ] [torch.randn()](https://pytorch.org/docs/stable/generated/torch.randn.html#torch.randn)

- [ ] [torch.ones()](https://pytorch.org/docs/stable/generated/torch.ones.html#torch.ones)

- [ ] [torch.tensor()](https://pytorch.org/docs/stable/generated/torch.tensor.html#torch.tensor)

- [ ] [torch.zeros()](https://pytorch.org/docs/stable/generated/torch.zeros.html#torch.zeros)

### [Named dimensions](https://pytorch.org/docs/stable/named_tensor.html#named-dimensions)

**Comment by GeorgeS2019** [comment]
FYI: T[orchScript Support for Named Tensors](https://github.com/pytorch/pytorch/issues/36603)

[Other supported (and unsupported) features](https://pytorch.org/tutorials/intermediate/named_tensor_tutorial.html#other-supported-and-unsupported-features)
- [ ] JIT support; (Not Yet, perhaps in 1.3)

**Pull request merged**